nesta/core/tasks/projects/ai_diversity/doc2cluster.py

```diff
@@ -0,0 +1,227 @@
+    predictions from the DB.
+- end: Exit metaflow pipeline.
+
+The pipeline takes a set of parameters. Two examples:
+1. Run the pipeline and fit a new HDBSCAN model
+python doc2cluster.py --no-pylint run --transformer distilbert-base-nli-stsb-mean-tokens --db_config mysqldb.config --min_cluster_size 10 --min_samples 2 --new_clusterer True --clusterer_name hdbscan_model --s3_bucket test-document-vectors
+
+2. Run the pipeline and predict a cluster distribution for vectors using a fitted HDBSCAN model 
+python doc2cluster.py --no-pylint run --transformer distilbert-base-nli-stsb-mean-tokens --db_config mysqldb.config --new_clusterer False --clusterer_name hdbscan_model --s3_bucket test-document-vectors
+"""
+
+from metaflow import FlowSpec, step, Parameter
+from sqlalchemy import create_engine
+from sqlalchemy.sql import exists
+from sqlalchemy.engine.url import URL
+from sqlalchemy.orm import sessionmaker
+import logging
+from sentence_transformers import SentenceTransformer
+import hdbscan
+import numpy as np
+from nesta.core.luigihacks import misctools
+from nesta.core.orms.arxiv_orm import Article, ArticleVector, ArticleCluster
+from nesta.packages.misc_utils.s3_utils import store_on_s3, load_from_s3
+
+
+class Doc2ClusterFlow(FlowSpec):
+    """Fetches paper abtracts from a SQL database, transforms them to vectors
+    and clusters them.
+    """
+
+    db_config = Parameter(
+        "db_config", help="DB configuration filename", required=True, type=str
+    )
+    bert_model = Parameter(
+        "transformer",
+        help="""
+        Name of the model used to encode documents.
+        Full list: https://docs.google.com/spreadsheets/d/14QplCdTCDwEmTqrn1LH4yrbKvdogK4oQvYO1K1aPR5M/edit?usp=sharing
+        """,
+        required=True,
+        type=str,
+    )
+    new_clusterer = Parameter(
+        "new_clusterer",
+        help="Fits a new cluster model using all the document vectors. If set to False, it will fetch a fitted cluster model from S3.",
+        required=True,
+        type=bool,
+    )
+    clusterer_name = Parameter(
+        "clusterer_name", help="Filename of the clusterer.", required=True, type=str
+    )
+    min_cluster_size = Parameter(
+        "min_cluster_size",
+        help="HDBSCAN: the smallest size grouping that you wish to consider a cluster. Required only when new_clusterer=True.",
+        required=False,
+        type=int,
+    )
+    min_samples = Parameter(
+        "min_samples",
+        help="""HDBSCAN: Measure of how conservative you want the clustering to be.
+        The larger the value of min_samples you provide, the more conservative the clustering. Required only when new_clusterer=True.""",
+        required=False,
+        type=int,
+    )
+    s3_bucket = Parameter(
+        "s3_bucket",
+        help="S3 bucket to store/load the clusterer.",
+        required=True,
+        type=str,
+    )
+
+    def _create_db_session(self, config, header="client", database="production"):
```

Comment:
could you switch to the `orm_utils.db_session` tool, see usage example here:

https://github.com/nestauk/nesta/blob/dev/nesta/core/batchables/arxiv/arxiv_elasticsearch/run.py#L102

asides from avoiding code repetition, it also has the commit/rollback built-in, plus it will wait until a connection is permitted, if the mysql db is busy.